File: erlang/ch1/ex123.erl
%%%-------------------------------------------------------------------
%%% @author Yohan Lee
%%% @copyright (C) 2019
%%% @doc
%%% cd("c:/Users/Yohan/IdeaProjects/sicp/erlang/ch1").
%%% compile with > c("exXXX.erl").
%%% if already compiled, load with > l(exXXX).
%%% run with > exXXX:main().
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(ex123).
-author("Yohan").

%% API
-export([main/0, smallest_divisor/1]).

main() ->
  io:fwrite("~w\n", [smallest_divisor(199)]),
  io:fwrite("~w\n", [smallest_divisor(1999)]),
  io:fwrite("~w\n", [smallest_divisor(19999)]),
  io:fwrite("~w\n", [smallest_divisor(15)]).


smallest_divisor(N) -> find_divisor(N, 2).

find_divisor(N, Test) when Test * Test > N -> N;
find_divisor(N, Test) ->
  if
    N rem Test == 0 -> Test;
    true -> find_divisor(N, Test + 1)
  end.

next(2) -> 3;
next(N) -> N + 2.

File: elixir/ch1/ex123.ex
# cd("c:/Users/Yohan/IdeaProjects/sicp/elixir/ch1")
# Load with import_file("filename")
# Run with Module.func(args)
defmodule Ex123 do
  @moduledoc false

  def sq(x), do: x * x

  def divides?(a, b), do: rem(b, a) == 0

  def smallest_divisor(n), do: find_divisor(n, 2)

  def next(2), do: 3
  def next(n), do: n + 2

  def find_divisor(n, test_divisor) when test_divisor * test_divisor > n, do: n
  def find_divisor(n, test_divisor) do
    if divides?(test_divisor, n) do
      test_divisor
    else
      find_divisor(n, next(test_divisor))
    end
  end

end
